### PR TITLE
Expand type support for random uniform() & randint()

### PR DESCRIPTION
**Pull request by jakevdp**
Adds `float16` and `bfloat16` support to `uniform()`, and `(u)int8` & `(u)int16` support to `randInt`

### CODE REVIEW

**Comment by jakevdp** [comment]
TPU tests still failing.

File: tests/random_test.py [hunk 1]
```diff
@@ -158,8 +164,10 @@ def testRngRandomBits(self):
 
   @parameterized.named_parameters(jtu.cases_from_list(
       {"testcase_name": "_{}".format(dtype), "dtype": np.dtype(dtype).name}
-      for dtype in [np.float32, np.float64]))
+      for dtype in float_dtypes))
   def testRngUniform(self, dtype):
+    if jtu.device_under_test() == "tpu" and jnp.dtype(dtype).itemsize < 3:
+      raise SkipTest("rng.uniform() not supported on TPU for 16-bit types.")
```
Comment (jekbradbury) [inline tests/random_test.py:170]:
```suggestion
      raise SkipTest("random.uniform() not supported on TPU for 16-bit types.")
```

File: tests/random_test.py [hunk 2]
```diff
@@ -173,8 +181,10 @@ def testRngUniform(self, dtype):
 
   @parameterized.named_parameters(jtu.cases_from_list(
       {"testcase_name": "_{}".format(dtype), "dtype": np.dtype(dtype).name}
-      for dtype in [np.int32, np.int64]))
+      for dtype in int_dtypes + uint_dtypes))
   def testRngRandint(self, dtype):
+    if jtu.device_under_test() == "tpu" and jnp.dtype(dtype).itemsize < 3:
+      raise SkipTest("rng.randint() not supported on TPU for 8- or 16-bit types.")
```
Comment (jekbradbury) [inline tests/random_test.py:187]:
```suggestion
      raise SkipTest("random.randint() not supported on TPU for 8- or 16-bit types.")
```

File: jax/dtypes.py [hunk 3]
```diff
@@ -120,7 +122,8 @@ def coerce_to_array(x):
 def finfo(dtype):
   # Since NumPy doesn't consider bfloat16 a floating-point type, we have to
   # provide an alternative implementation of finfo that does so.
-  if np.result_type(dtype) == _bfloat16_dtype:
+  if ((isinstance(dtype, str) and dtype == "bfloat16") or
+      np.result_type(dtype) == _bfloat16_dtype):
```
Comment (jekbradbury) [inline jax/dtypes.py:126]: Are these two different kinds of bfloat16 types?

File: jax/dtypes.py [hunk 4]
```diff
@@ -77,6 +77,8 @@ class _bfloat16_finfo(object):
 @util.memoize
 def canonicalize_dtype(dtype):
   """Convert from a dtype to a canonical dtype based on FLAGS.jax_enable_x64."""
+  if isinstance(dtype, str) and dtype == "bfloat16":
+    dtype = bfloat16
```
Comment (jekbradbury) [inline jax/dtypes.py:81]: What's changed that makes this needed?

File: jax/dtypes.py (hunk 3 above)
Comment (jakevdp) [reply jax/dtypes.py:126]: This update lets the function handle ``"bfloat16"`` as a string. ``np.result_type`` accepts string inputs, but doesn't know about bfloat16.

File: jax/dtypes.py (hunk 4 above)
Comment (jakevdp) [reply jax/dtypes.py:81]: Passing all float dtypes to ``testRngUniform`` leads to an error for bfloat16 without this change.

File: tests/random_test.py (hunk 2 above)
Comment (jakevdp) [reply tests/random_test.py]: Done.

File: tests/random_test.py (hunk 1 above)
Comment (jakevdp) [reply tests/random_test.py]: Done.

**Review by jekbradbury — APPROVED** [review]

**Comment by jakevdp** [comment]
This broke TPU tests; fix in #3153

**Pull request merged**